EmployeeList/ViewModels/CustomImageViewModel.swift

```diff
@@ -0,0 +1,142 @@
+    
+    @Published var isErrorShown = false
+    @Published var errorMessage = ""
+    
+    init(imageService: ImageServiceType = ImageService(), imageURL: String) {
+        self.imageService = imageService
+        self.imageURL = imageURL
+        
+        bindInputsToOutputs()
+        print(" --- custom image view model init done ---")
+    }
+    
+    // Publish values on-demand by calling the send() method
+    func update() {
+        onAppearSubject.send(())
+    }
+    
+    private func buildImageKey(_ url: String) -> String {
+        if let imageURLComponents = URL(string: url)?.pathComponents {
+            if imageURLComponents.count == 1 {
+                return url
+            }
+            let endIndex = imageURLComponents.endIndex
+            let id = imageURLComponents[endIndex - 2]
+            let size = imageURLComponents[endIndex - 1]
+            return id + size
+        }
+        return url
+    }
+    
+    func bindInputsToOutputs() {
+        let imageKey = buildImageKey(self.imageURL)
+
+        if self.imageService.existsInDisk(imageKey) {
+            print("Load image from disk")
+            loadFromDisk(from: imageKey)
+            loadImage()
+        } else {
+            print("Load image from web")
+            loadFromWeb(from: self.imageURL)
+            loadImage()
+        }
+
+    }
+    
+    
+    // MARK: Bind inputs
+    func loadFromDisk(from key: String) {
+
+        let publisher = onAppearSubject.flatMap { _ ->  AnyPublisher<Data, Error> in
+            return self.imageService.loadFromDisk(forKey: key)
+        }
+        let diskStream = publisher.subscribe(imageSubject)
+        
+        cancellables["disk"] = diskStream
+    }
+    
+    func loadFromWeb(from url: String) {
+        
+        let endpoint = ImageRequest(path: url)
+        let publisher = onAppearSubject.flatMap { _ -> AnyPublisher<Data, Error> in
+            return self.imageService.call(from: endpoint)
+        }
+        let webStream = publisher.subscribe(imageSubject)
+        cancellables["web"] = webStream
+    }
+    
+    // MARK: Only save the image to disk the first time the image is downloaded from the web
+    func saveToDisk(_ image: UIImage, _ key: String) {
+
+        let publisher = diskSaveSubject.flatMap { _ -> AnyPublisher<Bool, Error> in
+            return self.imageService.saveToDisk(self.image, forKey: key)
+        }
+        
+        let diskSaveStream = publisher
+            .map { $0 }
+            .sink(receiveCompletion: { result in
+                switch result {
+                case .failure(let error): do {
+                    self.isErrorShown = true
+                    self.errorMessage = error.localizedDescription
+                    }
+                case .finished:
+                    break
+                }
+            }, receiveValue: { data in
+                print("Successfully stored image, boolean is \(data)")
```

Comment:
are you gonna do anything with the data?

Reply:
I expect the data here is a boolean indicating if saving an image to disk succeeds or not. But actually, an error is diplayed if any error is thrown during storing procudure. Error is handled by
` case .failure(let error): do {
                    self.isErrorShown = true
                    self.errorMessage = error.localizedDescription
                    }`

If all things are good, data is always `true`.

I feel like this may not be good, but don't know how to improve it.